stacks/api/machinery/src/store.ts

```diff
@@ -0,0 +1,115 @@
+import {DatabaseService, BaseCollection} from "@spica-server/database";
+import {Resource} from "./definition";
+import {GroupResource} from "./scheme";
+
+let db: DatabaseService;
+
+/**
+ * Do not use this function if you do not know what you are doing.
+ */
+export function __setDb(databaseService: DatabaseService) {
+  db = databaseService;
+}
+
+class ObjectStore extends BaseCollection<Resource<unknown, unknown>>("objects") {}
+
+type DeepPartial<T> = {
+  [P in keyof T]?: DeepPartial<T[P]>;
+};
+
+export function store(groupResource: GroupResource) {
+  return new Store(groupResource);
+}
+
+export class Store<SpecType = unknown, StatusType = unknown> {
+  private store = new ObjectStore(db);
+  private groupKey: string;
+
+  constructor(groupResource: GroupResource) {
+    this.groupKey = `${groupResource.group}ɵ${groupResource.resource}`;
+  }
+
+  async get(name: string): Promise<Resource<SpecType, StatusType>> {
+    const result = (await this.store.findOne({
+      _id: `${this.groupKey}ɵ${name}`
+    })) as Resource<SpecType, StatusType>;
+    if (result) {
+      // it is there. i know it
+      delete result["_id"];
+    }
+    return result;
+  }
+
+  async set(name: string, value: Resource<SpecType, StatusType>): Promise<void> {
+    value["_id"] = `${this.groupKey}ɵ${name}`;
+    await this.store.replaceOne({_id: value["_id"]}, value, {
+      upsert: true
+    });
+  }
+
+  async patch(name: string, patch: DeepPartial<Resource<SpecType, StatusType>>) {
```

Comment:
We have the same method on bucket utilities. Can we move that `getUpdateQueryForPatch` method to the common package and use it?

Reply:
This an implementation detail of the store functionality of machinery. TBH I don't wanna tie it to a common function since this might change in the future. I think for now we better not couple it with any other patching mechanism in the codebase.

Reply:
This might change in the future with an extended version of merge+patch that we are going to implement.